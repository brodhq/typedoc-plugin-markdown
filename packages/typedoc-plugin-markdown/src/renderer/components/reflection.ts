import { DeclarationReflection } from 'typedoc';

import { MarkdownBuilder } from '../../markdown-tools//builder';
import { heading } from '../../markdown-tools//elements';
import { CommentsComponent } from './comments.component';
import { GroupComponent } from './group';
import { HierarchyComponent } from './hierachy';
import { ImplementsComponent } from './implements';
import { TocComponent } from './toc.component';
import { TypeParameterTableComponent } from './type-parameter-table';

export function ReflectionComponent(model: DeclarationReflection) {
  const md = new MarkdownBuilder();

  if (model.comment) {
    md.add(CommentsComponent(model.comment));
  }

  if (model.typeParameters) {
    md.add(heading(2, 'Type parameters'));
    md.add(TypeParameterTableComponent(model.typeParameters));
  }

  if (model.typeHierarchy && model.typeHierarchy.next) {
    md.add(heading(2, 'Hierarchy'));
    md.add(HierarchyComponent(model.typeHierarchy));
  }

  if (model.implementedTypes) {
    md.add(heading(2, 'Implements'));
    md.add(ImplementsComponent(model.implementedTypes));
  }

  if (model.signatures) {
    md.add(heading(2, 'Callable'));
    model.signatures.forEach((signature) => {
      md.add(heading(3, signature.name));
    });
  }

  md.add(TocComponent(model));

  model.groups
    ?.filter((group) => !group.allChildrenHaveOwnDocument())
    .forEach((group) => {
      md.add(GroupComponent(group));
    });

  return md.print();
}
