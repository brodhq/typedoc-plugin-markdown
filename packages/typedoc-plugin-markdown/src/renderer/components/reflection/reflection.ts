import { DeclarationReflection } from 'typedoc';

import { MarkdownBuilder } from '../../../markdown-tools/builder';
import { heading, unorderedList } from '../../../markdown-tools/elements';
import { SignatureComponent } from '../member/member.signature';
import { GroupsComponent } from '../page/groups';
import { TocComponent } from '../page/toc';
import { CommentsComponent } from '../shared/comments';
import { TypeParameterTableComponent } from '../shared/table.type-parameter';
import { TypeComponent } from '../shared/type';
import { HierarchyComponent } from './reflection.hierachy';
import { ReflectionPathComponent } from './reflection.path';

export function ReflectionComponent(model: DeclarationReflection) {
  const md = new MarkdownBuilder();

  md.add(ReflectionPathComponent(model));

  if (model.comment) {
    md.add(CommentsComponent(model.comment));
  }

  if (model.typeParameters) {
    md.add(heading(2, 'Type parameters'));
    md.add(TypeParameterTableComponent(model.typeParameters));
  }

  if (model.typeHierarchy && model.typeHierarchy.next) {
    md.add(heading(2, 'Hierarchy'));
    md.add(HierarchyComponent(model.typeHierarchy));
  }

  if (model.implementedTypes) {
    md.add(heading(2, 'Implements'));
    md.add(
      unorderedList(
        model.implementedTypes.map((implementedType) =>
          TypeComponent(implementedType),
        ),
      ),
    );
  }

  if (model.signatures) {
    md.add(heading(2, 'Callable'));
    model.signatures.forEach((signature) => {
      md.add(heading(3, signature.name));
      md.add(SignatureComponent(signature));
    });
  }

  md.add(TocComponent(model));

  if (model.groups) {
    md.add(GroupsComponent(model.groups));
  }

  return md.print();
}
