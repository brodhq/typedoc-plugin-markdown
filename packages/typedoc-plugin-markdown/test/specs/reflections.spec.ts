import * as groupsComponent from '../../src/renderer/components/page/groups';
import * as tocComponent from '../../src/renderer/components/page/toc';
import { ReflectionComponent } from '../../src/renderer/components/reflection/reflection';
import { TestApp } from '../test-app';

describe(`Reflections:`, () => {
  let testApp: TestApp;

  beforeAll(async () => {
    jest.spyOn(tocComponent, 'TocComponent').mockReturnValue('[TOC]');
    jest.spyOn(groupsComponent, 'GroupsComponent').mockReturnValue('[GROUPS]');
    testApp = new TestApp(['reflections.ts', 'hierarchy.ts']);
    await testApp.bootstrap();
  });

  describe(`(template)`, () => {
    test(`should compile a callable reflection`, () => {
      const reflection = testApp.findReflectionByName('CallableReflection');
      expect(ReflectionComponent(reflection)).toMatchSnapshot();
    });

    test(`should compile an indexable reflection`, () => {
      const reflection = testApp.findReflectionByName('IndexableReflection');
      expect(ReflectionComponent(reflection)).toMatchSnapshot();
    });

    test(`should compile implemented class`, () => {
      const reflection = testApp.findReflectionByName('ImplementedClass');
      expect(ReflectionComponent(reflection)).toMatchSnapshot();
    });

    test(`should compile Enum`, () => {
      const reflection = testApp.findReflectionByName('EnumReflection');
      expect(ReflectionComponent(reflection)).toMatchSnapshot();
    });
  });
  describe(`(hierarchy)`, () => {
    test(`should compile type hierarchy`, () => {
      const reflection = testApp.findReflectionByName('ParentClass');
      expect(ReflectionComponent(reflection)).toMatchSnapshot();
    });

    test(`should compile nested type hierarchy`, () => {
      const reflection = testApp.findReflectionByName('ChildClassA');
      expect(ReflectionComponent(reflection)).toMatchSnapshot();
    });
  });
});
