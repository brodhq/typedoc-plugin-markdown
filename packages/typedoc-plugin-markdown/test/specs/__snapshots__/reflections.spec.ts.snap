// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reflections: (hierarchy) should compile nested type hierarchy 1`] = `
"## Hierarchy

- [\`ParentClass\`](classes/parentclass.md)

  ↳ **\`ChildClassA\`**

  ↳↳ [\`GrandChildClassA\`](classes/grandchildclassa.md)

[TOC]

[GROUPS]"
`;

exports[`Reflections: (hierarchy) should compile type hierarchy 1`] = `
"## Hierarchy

- **\`ParentClass\`**

  ↳ [\`ChildClassA\`](classes/childclassa.md)

  ↳ [\`ChildClassB\`](classes/childclassb.md)

[TOC]

[GROUPS]"
`;

exports[`Reflections: (template) should compile Enum 1`] = `
"[TOC]

[GROUPS]"
`;

exports[`Reflections: (template) should compile a callable reflection 1`] = `
"## Callable

### CallableReflection

▸ **CallableReflection**(): \`string\`

#### Returns

\`string\`

#### Defined in

[reflections.ts:6](https://github.com/tgreyuk/typedoc-plugin-markdown/blob/a1a25ce/packages/typedoc-plugin-markdown/test/stubs/src/reflections.ts#L6)

[TOC]"
`;

exports[`Reflections: (template) should compile an indexable reflection 1`] = `"[TOC]"`;

exports[`Reflections: (template) should compile implemented class 1`] = `
"## Implements

- [\`ReflectionClass\`](classes/reflectionclass.md)

[TOC]

[GROUPS]"
`;
